Foundation/NSURLSession/EasyHandle.swift

```diff
@@ -0,0 +1,632 @@
+        static let sendPaused = _PauseState(rawValue: 1 << 1)
+    }
+}
+extension URLSessionTask._EasyHandle._PauseState {
+    func setState(on handle: URLSessionTask._EasyHandle) {
+        try! CFURLSessionEasyHandleSetPauseState(handle.rawHandle, contains(.sendPaused) ? 1 : 0, contains(.receivePaused) ? 1 : 0).asError()
+    }
+}
+extension URLSessionTask._EasyHandle._PauseState : Streamable {
+    func write<Target : OutputStream>(to target: inout Target) {
+        switch (self.contains(.receivePaused), self.contains(.sendPaused)) {
+        case (false, false): target.write("unpaused")
+        case (true, false): target.write("receive paused")
+        case (false, true): target.write("send paused")
+        case (true, true): target.write("send & receive paused")
+        }
+    }
+}
+extension URLSessionTask._EasyHandle {
+    /// Pause receiving data.
+    ///
+    /// - SeeAlso: https://curl.haxx.se/libcurl/c/curl_easy_pause.html
+    func pauseReceive() {
+        URLSession.printDebug("[EasyHandle] pause receive (\(pauseState))")
+        guard !pauseState.contains(.receivePaused) else { return }
+        pauseState.insert(.receivePaused)
+        pauseState.setState(on: self)
+    }
+    /// Pause receiving data.
+    ///
+    /// - Note: Chances are high that delegate callbacks (with pending data)
+    /// will be called before this method returns.
+    /// - SeeAlso: https://curl.haxx.se/libcurl/c/curl_easy_pause.html
+    func unpauseReceive() {
+        URLSession.printDebug("[EasyHandle] unpause receive (\(pauseState))")
+        guard pauseState.contains(.receivePaused) else { return }
+        pauseState.remove(.receivePaused)
+        pauseState.setState(on: self)
+    }
+    /// Pause sending data.
+    ///
+    /// - SeeAlso: https://curl.haxx.se/libcurl/c/curl_easy_pause.html
+    func pauseSend() {
+        URLSession.printDebug("[EasyHandle] pause send (\(pauseState))")
+        guard !pauseState.contains(.sendPaused) else { return }
+        pauseState.insert(.sendPaused)
+        pauseState.setState(on: self)
+    }
+    /// Pause sending data.
+    ///
+    /// - Note: Chances are high that delegate callbacks (with pending data)
+    /// will be called before this method returns.
+    /// - SeeAlso: https://curl.haxx.se/libcurl/c/curl_easy_pause.html
+    func unpauseSend() {
+        URLSession.printDebug("[EasyHandle] unpause send (\(pauseState))")
+        guard pauseState.contains(.sendPaused) else { return }
+        pauseState.remove(.sendPaused)
+        pauseState.setState(on: self)
+    }
+}
+
+internal extension URLSessionTask._EasyHandle {
+    /// errno number from last connect failure
+    /// - SeeAlso: https://curl.haxx.se/libcurl/c/CURLINFO_OS_ERRNO.html
+    var connectFailureErrno: CInt {
+        var errno = Int()
+        try! CFURLSession_easy_getinfo_long(rawHandle, CFURLSessionInfoOS_ERRNO, &errno).asError()
+        return CInt(errno)
```

Comment:
We should strive to make the C interface disappear as soon as possible when importing into Swift - so API like this should probably return `Int` instead of something like `CInt`.